Update docs for pagination. (#1592)
opensearch-project/sql: docs/dev/Pagination-v2.md

# REST API
## Initial Query Request

> Initial query request contains the search request and page size. It can't be changed later while scrolling through pages issued by this request. Search query to OpenSearch engine is built during processing the initial request.

Author: Added in 3a4f808c1.

`cursor_id` will be different with each request.

## End of scrolling/paging

> When scrolling is finished, SQL plugin still returns a cursor. This cursor leads to the final page, which as no cursor and no data. Receiving that page means all data was properly queried and returned to user. Cursor is closed automatically on processing that page.

Reviewer: ```suggestion
When scrolling is finished, SQL plugin returns a final cursor. This cursor leads to an empty page, which has no cursor and no data hits. Receiving that page means all data was properly queried, and the scrolling cursor has been closed.
```

The client will receive an [error response](#error-response) if executing this request results in an OpenSearch or SQL plug-in error.

## Cursor Keep Alive Timeout

Each cursor has a keep alive timer associated with it. When the timer runs out, the cursor is automatically closed by OpenSearch.

This timer is reset every time a page is retrieved.

The client will receive an [error response](#error-response) if it sends a cursor request for an expired cursor.

> Keep alive timeout is [configurable](../user/admin/settings.rst#plugins.sql.cursor.keep_alive) by setting `plugins.sql.cursor.keep_alive` and has default value 1 minute.

Author: Added in 3a4f808c1.

## Error Response

The client will receive an error response if any of the above REST calls result in an server-side error.

The response object has the following format:
```json
{
    "error": {
        "details": "<string>",
        "reason": "<string>",
        "type": "<string>"
    },

>     "status": "<integer>"

Reviewer: I know this satisfies the json format, but this looks wrong

Author: Fixed in d3888ceaf

## SQL Node Load Balancing
V2 SQL engine supports *sql node load balancing* -- a cursor request can be routed to any SQL node in a cluster. This is achieved by encoding all data necessary to retrieve the next page in the `cursor_id`.

## Feature Design

> 

Reviewer: One thing that is not defined in this section is "How do we determine what is a paged request vs a non-paged request?".  We should say where it's checked and what it checks to make this determination.

Author: Fixed in d3888ceaf

### Plan Tree changes

> There are different plan trees are built during request processing. Read more about their purpose and stages [here](query-optimizer-improvement.md#Examples). Article below describes what changes are introduced in these trees by pagination feature.

Reviewer: ```suggestion
Different plan trees are built during request processing. Read more about their purpose and stages [here](query-optimizer-improvement.md#Examples). The section below describes the changes being introduced to these trees by the pagination feature.
```

> Simplified workflow of plan trees is shown below. Initial Page Request is processed the same way as non-paging request.

Reviewer: ```suggestion
Simplified workflow of plan trees is shown below. Initial Page Request is processed the same way as a non-paging request.
```

Reviewer: > Initial Page Request is processed the same way as non-paging request
Is this true?  Why give us two diagrams for "Non Paged Request" and "Initial Page Request"?  Why not just combine them into a single state diagram?

    state "Optimized Logical Plan Tree" as Optimized
    state "Physical Plan Tree" as Physical

    [*] --> Parse : ANTLR
    Parse --> Unresolved : AstBuilder
    Unresolved --> Abstract : QueryPlanner
    Abstract --> Logical : Planner
    Logical --> Optimized : Optimizer
    Optimized --> Physical : Implementor
  }
```
```mermaid
stateDiagram-v2
  state "Initial Page Request" as Paged {
    direction LR
    state "Parse Tree" as Parse
    state "Unresolved Plan Tree" as Unresolved
    state "Abstract Plan Tree" as Abstract
    state "Logical Plan Tree" as Logical
    state "Optimized Logical Plan Tree" as Optimized
    state "Physical Plan Tree" as Physical

    [*] --> Parse : ANTLR
    Parse --> Unresolved : AstBuilder
    Unresolved --> Abstract : QueryPlanner
    Abstract --> Logical : Planner
    Logical --> Optimized : Optimizer
    Optimized --> Physical : Implementor
  }
```
```mermaid
stateDiagram-v2
  state "Subsequent Page Request" as Paged {
    direction LR
    state "Abstract Plan Tree" as Abstract
    state "Physical Plan Tree" as Physical

    [*] --> Abstract : QueryPlanner
    Abstract --> Physical : Deserializer
  }
```

> New plan tree workflow was added for Subsequent Page Requests. Since a final Physical Plan tree was already created for Initial request, subsequent requests should have the same tree. The tree is being serialized into `cursor` by `PlanSerializer` and then restored back. Query parsing and analysis is also not performed for cursor requests, so these steps are also not executed there.

Reviewer: ```suggestion
New plan tree workflow was added for Subsequent Page Requests. Since a final Physical Plan tree was already created for Initial request, subsequent requests should have the same tree. The tree is serialized into a `cursor` by `PlanSerializer` to be de-serialized on the subsequence Page Request. Query parsing and analysis is not performed for Subsequent Page Requests, since the Physical Plan tree is instead de-serialized from the `cursor`.
```

#### Abstract Plan tree

> Abstract Plan Tree for non-paged requests remains unchanged by pagination feature. `QueryPlan` as a top entry  got new optional field `pageSize`, which is unset for non-paged requests.

Reviewer: ```suggestion
Abstract Plan Tree for non-paged requests remains unchanged. The `QueryPlan`, as a top entry`, has a new optional field `pageSize`, which is not defined for non-paged requests. 
```

```mermaid
classDiagram
  direction LR
  class QueryPlan {
    <<AbstractPlan>>
    -Optional~int~ pageSize
    -UnresolvedPlan plan
    -QueryService queryService
  }
  class UnresolvedPlanTree {
    <<UnresolvedPlan>>
  }
  QueryPlan --* UnresolvedPlanTree
```

Abstract plan tree for Initial Query Request has following changes:
1. New Plan node -- `Paginate` -- added into the tree.
2. `pageSize` parameter in `QueryPlan` is set, and `Paginate` is being added. It is converted to `LogicalPaginate` later.

```mermaid
classDiagram
  direction LR
  class QueryPlan {
    <<AbstractPlan>>
    -Optional~int~ pageSize
    -UnresolvedPlan plan
    -QueryService queryService
  }
  class Paginate {
    <<UnresolvedPlan>>
    -int pageSize
    -UnresolvedPlan child
  }
  class UnresolvedPlanTree {
    <<UnresolvedPlan>>
  }
  QueryPlan --* Paginate
  Paginate --* UnresolvedPlanTree
```

> Non-paging requests have the same plan tree, but `pageSize` value in `QueryPlan` is unset.

Reviewer: Is it necessary to say this twice?

Author: I can't find the second time, where is it?

The examples below show Abstract Plan Tree for the same query in different request types:

```mermaid
stateDiagram-v2
  state "Non Paged Request" as NonPaged {
    state "QueryPlan" as QueryPlanNP
    state "Project" as ProjectNP
    state "Limit" as LimitNP
    state "Filter" as FilterNP
    state "Aggregation" as AggregationNP
    state "Relation" as RelationNP

    QueryPlanNP --> ProjectNP
    ProjectNP --> LimitNP
    LimitNP --> FilterNP
    FilterNP --> AggregationNP
    AggregationNP --> RelationNP
  }

  state "Initial Query Request" as Paged {
    state "QueryPlan" as QueryPlanIP
    state "Project" as ProjectIP
    state "Limit" as LimitIP
    state "Filter" as FilterIP
    state "Aggregation" as AggregationIP
    state "Relation" as RelationIP

    Paginate --> QueryPlanIP
    QueryPlanIP --> ProjectIP
    ProjectIP --> LimitIP
    LimitIP --> FilterIP
    FilterIP --> AggregationIP
    AggregationIP --> RelationIP
  }

  state "Subsequent Query Request" As Sub {
    ContinuePaginatedPlan
  }
```

#### Logical Plan tree

Changes to plan tree for Initial Query Request with pagination:

> 1. `LogicalPaginate` is added to the top of the tree. It stores information about paging/scrolling should be done in a private field `pageSize` being pushed down in the `Optimizer`.

Reviewer: What is the `1.` for?

Author: There is a list of changes. Luckly, it has only one entry.

```mermaid
classDiagram
  direction LR
  class LogicalProject {
    <<LogicalPlan>>
  }
  class OpenSearchPagedIndexScanBuilder {
    <<TableScanBuilder>>
  }

  LogicalProject --* OpenSearchPagedIndexScanBuilder
```

Note: No Logical Plan tree created for Subsequent Query Request.

The examples below show optimized Logical Plan Tree for the same query in different request types:

```mermaid
stateDiagram-v2
  state "Non Paged Request" as NonPaged {
    state "LogicalProject" as ProjectNP
    state "LogicalLimit" as LimitNP
    state "LogicalSort" as SortNP
    state "OpenSearchIndexScanQueryBuilder" as RelationNP

    ProjectNP --> LimitNP
    LimitNP --> SortNP
    SortNP --> RelationNP
  }

  state "Initial Paged Request" as Paged {
    state "LogicalProject" as ProjectIP
    state "LogicalLimit" as LimitIP
    state "LogicalSort" as SortIP
    state "OpenSearchPagedIndexScanBuilder" as RelationIP

    ProjectIP --> LimitIP
    LimitIP --> SortIP
    SortIP --> RelationIP
  }
```

#### Physical Plan tree

Changes:
1. `OpenSearchPagedIndexScanBuilder` is converted to `OpenSearchPagedIndexScan` by `Implementor`.

> 2. Entire Physical Plan tree is created by `PlanSerializer` for Subsequent Query requests. The deserialized tree has the same structure which Initial Query Request had.

Reviewer: ```suggestion
2. Entire Physical Plan tree is created by `PlanSerializer` for Subsequent Query requests. The deserialized tree has the same structure as the Initial Query Request.
```

    <<TableScanOperator>>
  }

  ProjectOperator --* OpenSearchPagedIndexScan
```

The examples below show Physical Plan Tree for the same query in different request types:

```mermaid
stateDiagram-v2
  state "Non Paged Request" as NonPaged {
    state "ProjectOperator" as ProjectNP
    state "LimitOperator" as LimitNP
    state "SortOperator" as SortNP
    state "OpenSearchIndexScan" as RelationNP

    ProjectNP --> LimitNP
    LimitNP --> SortNP
    SortNP --> RelationNP
  }

  state "Initial Query Request" as Paged {
    state "ProjectOperator" as ProjectIP
    state "LimitOperator" as LimitIP
    state "SortOperator" as SortIP
    state "OpenSearchPagedIndexScan" as RelationIP

    ProjectIP --> LimitIP
    LimitIP --> SortIP
    SortIP --> RelationIP
  }

  state "Subsequent Query Request" As Sub {
    state "ProjectOperator" as ProjectSP
    state "LimitOperator" as LimitSP
    state "SortOperator" as SortSP
    state "OpenSearchPagedIndexScan" as RelationSP

    ProjectSP --> LimitSP
    LimitSP --> SortSP
    SortSP --> RelationSP
  }
```

### Architecture Diagrams

New code workflows which added by Pagination feature are highlighted.

#### Non Paging Query Request

> To simplify comparison and understanding, below represented processing a regular query request. This workflow wasn't changed by Pagination feature.

Reviewer: ```suggestion
A non-paging request sequence diagram is shown below for comparison: 
```

    ResourceMonitorPlan-->>-ProjectOperator: value
  ProjectOperator-->>-OpenSearchExecutionEngine: value
```

> ### Plan Tree changes

Reviewer: A couple recommended changes from above in the document: 
* Under #Scope, I think we can be more specific about what is out of scope vs what will be completed.  From what I understand: the document applies to all clauses (select...where...order by...) except for aggregation (group by).  Maybe some of those clauses aren't yet implemented (and that can be under the schedule) but the concepts still hold.  
* Fix: `The client receives an (error response](#error-response) if:` => `The client receives an [error response](#error-response) if:`
* Can we make the language consistent between initial page request and next page request:  `"cursor": /* cursor_id */` vs `"cursor": "<cursor_id>"`
* Maybe make section for `When Next Page Request is the Last Page` after `Next Page Request`?  And move the items that apply to the last page request to that section? 
* Please make the sub-categories under REST API and Design Diagrams consistent:  `Initial Query Request` vs `First Page` and `Next Page Request` vs `Second Page`.
* Under Design Diagrams/Second page: we include initial cursor verification in the sequence diagram.
* Do we need to update the `Total Hits` section for Peng's comments? 
*

> #### Abstract Plan tree

Reviewer: consider adding a state diagram to say both pagination and non-pagination both follow the following execution: 
```
<Abstract Plan> -> <Logical Plan> -> <Optimized Logical Plan> -> <Physical Plan> -> <Request>
```

We should seriously consider following the language used in https://user-images.githubusercontent.com/46505291/203645359-3f2fff73-a210-4bc0-a582-951a27de684d.jpg
* under docs/dev/query-optimizer-improvement.md Examples section.

> Changes:

Reviewer: Update to `Changes to plan tree for Initial Query Request with pagination` (or similar):
Add a section to `Changes without pagination`: no changes and pageSize will be empty/null.

Reviewer: We probably need a summary above for 'why' we need pageSize and the Paginate unresolved node.

1. New Plan node -- `Paginate` -- added into the tree.

> 2. `QueryPlan` got new optional field: page size. When it is set, `Paginate` is being added. It is converted to `LogicalPaginate` later. For non-paging requests the tree remains unchanged.

Reviewer: int is not an Optional.  Please consider making `pageSize` an Optional<Integer> instead of including an 'invalid' page size int.

```mermaid
classDiagram
  direction LR
  class QueryPlan {
    <<AbstractPlan>>
    -int pageSize
  }
  class Paginate {
    <<UnresolvedPlan>>
  }
  class UnresolvedPlanTree {
    <<UnresolvedPlan>>
  }
  QueryPlan --* Paginate
  Paginate --* UnresolvedPlanTree
```
#### Logical Plan tree

Changes:

> 1. `LogicalPaginate` is added to the top of the tree. It has a private field `pageSize` being pushed down in the `Optimizer`.

Reviewer: Following your template from above, consider adding: 
2. How is this different from a non-pagination logical plan?

Reviewer: Also include: what is the responsibility of the LogicalPaginate class for the LogicalPlan

```mermaid
classDiagram
  direction LR
  class LogicalPaginate {
    <<LogicalPlan>>
    -int pageSize
  }
  class LogicalPlanTree {
    <<LogicalPlan>>
  }
  class LogicalRelation {
    <<LogicalPlan>>
  }
  LogicalPaginate --* LogicalPlanTree
  LogicalPlanTree --* LogicalRelation
```

#### Optimized Plan tree

Changes:

> 1. There is a `OpenSearchPagedIndexScanBuilder` instead of `OpenSearchIndexScanQueryBuilder` on the bottom of the tree for paging requests. Non-paging requests are not affected. Both are instances of `TableScanBuilder` which extends `PhysicalPlan` interface.

Reviewer: ```suggestion
1. For pagination request, we push a `OpenSearchPagedIndexScanBuilder` instead of `OpenSearchIndexScanQueryBuilder` to the bottom of the tree. Both are instances of `TableScanBuilder` which extends `PhysicalPlan` interface.
```
You say how `OpenSearchPagedIndexScanBuilder` is similar to `OpenSearchIndexScanQueryBuilder`, but what is different?  What is the responsibility of `OpenSearchPagedIndexScanBuilder` in this state and why can't we re-use `OpenSearchIndexScanQueryBuilder`.  Justify your architecture changes.

2. `LogicalPaginate` is removed from the tree during push down operation in `Optimizer`.

See [article about `TableScanBuilder`](query-optimizer-improvement.md#TableScanBuilder) for more details.

```mermaid
classDiagram
  class LogicalProject {
    <<LogicalPlan>>
  }
  class OpenSearchPagedIndexScanBuilder {
    <<TableScanBuilder>>
  }

  LogicalProject --* OpenSearchPagedIndexScanBuilder
```

#### Physical Plan tree

Changes:

> 1. `OpenSearchPagedIndexScanBuilder` is converted to `OpenSearchPagedIndexScan` by `Implementor`.

Reviewer: Again:  why not just re-use `OpenSearchIndexScan`?  
Justify you architectural changes.

Author: They have different push down implementations. @MaxKsyunz is trying to merge them.
The doc shows the current stage of the things.

```mermaid
classDiagram
  direction LR
  class ProjectOperator {
    <<PhysicalPlan>>
  }
  class OpenSearchPagedIndexScan {

>     <<TableScanBuilder>>

Reviewer: I don't think this is a `*Builder` any longer

    participant CreatePagingTableScanBuilder
    participant OpenSearchExecutionEngine
    participant PlanSerializer

SQLService ->>+ QueryPlanFactory : execute

>   rect rgb(191, 223, 255)

Author: Change color, it is not readable on dark GH theme
![image](https://github.com/opensearch-project/sql/assets/88679692/c7f0ef52-a4bb-43e8-9a3e-1d3af8ed7fbc)

Author: Fixed in d3888ceaf

  QueryPlanFactory ->>+ CanPaginateVisitor : canConvertToCursor
    CanPaginateVisitor -->>- QueryPlanFactory : true
  end
  QueryPlanFactory ->>+ QueryService : execute
    QueryService ->>+ Planner : optimize
      rect rgb(191, 223, 255)
      Planner ->>+ CreatePagingTableScanBuilder : apply
        CreatePagingTableScanBuilder -->>- Planner : paged index scan
      end
      Planner -->>- QueryService : Logical Plan Tree
    QueryService ->>+ OpenSearchExecutionEngine : execute
      rect rgb(191, 223, 255)
      Note over OpenSearchExecutionEngine, PlanSerializer : Serialization
      OpenSearchExecutionEngine ->>+ PlanSerializer : convertToCursor
        PlanSerializer -->>- OpenSearchExecutionEngine : cursor
      end
      rect rgb(191, 223, 255)
      Note over OpenSearchExecutionEngine : get total hits
      end
      OpenSearchExecutionEngine -->>- QueryService : execution completed
    QueryService -->>- QueryPlanFactory : execution completed
  QueryPlanFactory -->>- SQLService : execution completed
```

#### Subsequent Query Request

> Second and further pages processed by a completely new workflow. The key point there:

Reviewer: ```suggestion
Subsequent pages are processed by a new workflow. The key point there:
```

sequenceDiagram
    participant SQLService
    participant QueryPlanFactory
    participant QueryService
    participant OpenSearchExecutionEngine
    participant PlanSerializer

SQLService ->>+ QueryPlanFactory : execute
  QueryPlanFactory ->>+ QueryService : execute
    rect rgb(191, 223, 255)
    note over QueryService, PlanSerializer : Deserialization
    QueryService ->>+ PlanSerializer: convertToPlan
      PlanSerializer -->>- QueryService: Physical plan tree
    end
    Note over QueryService : Planner, Optimizer and Implementor<br />are skipped
    QueryService ->>+ OpenSearchExecutionEngine : execute
      rect rgb(191, 223, 255)
      note over OpenSearchExecutionEngine, PlanSerializer : Serialization
      OpenSearchExecutionEngine ->>+ PlanSerializer : convertToCursor
        PlanSerializer -->>- OpenSearchExecutionEngine : cursor
      end
      rect rgb(191, 223, 255)
      Note over OpenSearchExecutionEngine : get total hits
      end
      OpenSearchExecutionEngine -->>- QueryService: execution completed
    QueryService -->>- QueryPlanFactory : execution completed
  QueryPlanFactory -->>- SQLService : execution completed
```

#### Legacy Engine Fallback

> Since pagination in V2 engine supports less SQL features than pagination in legacy engine, a fallback mechanism is created to keep V1 engine features still available for the end user. Pagination fallback is backed by a new exception type which allows legacy engine to intersect execution of a request.

Reviewer: ```suggestion
Since pagination in V2 engine supports fewer SQL commands than pagination in legacy engine, a fallback mechanism is created to keep V1 engine features still available for the end user. Pagination fallback is backed by a new exception type which allows legacy engine to intersect execution of a request.
```

    participant QueryPlanFactory
    participant CanPaginateVisitor

RestSQLQueryAction ->>+ SQLService : prepareRequest
  SQLService ->>+ QueryPlanFactory : execute
    rect rgb(191, 223, 255)
    note over SQLService, CanPaginateVisitor : V2 support check
    QueryPlanFactory ->>+ CanPaginateVisitor : canConvertToCursor
      CanPaginateVisitor -->>- QueryPlanFactory : false
    QueryPlanFactory -->>- RestSQLQueryAction : UnsupportedCursorRequestException
    deactivate SQLService
    end
      RestSQLQueryAction ->> Legacy Engine: accept
      Note over Legacy Engine : Processing in Legacy engine
        Legacy Engine -->> RestSQLQueryAction : complete
```

#### Serialization and Deserialization round trip

> The SQL engine should be able to completely recover the Physical Plan tree to continue its execution to get the next page. Serialization mechanism is responsible of proper dumping the plan tree. `ResourceMonitorPlan` shouldn't be serialized, because a new object of this type would be created for the restored plan tree before execution.

Reviewer: ```suggestion
The SQL engine should be able to completely recover the Physical Plan tree to continue its execution to get the next page. Serialization mechanism is responsible for recovering the plan tree. note: `ResourceMonitorPlan` isn't serialized, because a new object of this type would be created for the restored plan tree before execution. 
```

Serialization and Deserialization are performed by Java object serialization API.

> ## TODO describe ...

Reviewer: Please complete `TODO`

Author: Already completed, forgot to remove it.

        state "OpenSearchScrollRequest" as logState1_5
        logState1_1 --> logState1_2
        logState1_2 --> logState1_3
        logState1_3 --> logState1_4
        logState1_4 --> logState1_5
    }

    state "Deserialized Plan Tree" as SecondPageTree
    state SecondPageTree {
        state "ProjectOperator" as logState2_1
        state "..." as logState2_2
        state "OpenSearchPagedIndexScan" as logState2_3
        state "ContinuePageRequest" as logState2_4
        logState2_1 --> logState2_2
        logState2_2 --> logState2_3
        logState2_3 --> logState2_4
    }

    state "Subsequent Query Request Plan Tree" as SecondPage
    state SecondPage {
        state "ProjectOperator" as logState3_1
        state "..." as logState3_2
        state "ResourceMonitorPlan" as logState3_3
        state "OpenSearchPagedIndexScan" as logState3_4
        state "ContinuePageRequest" as logState3_5
        logState3_1 --> logState3_2
        logState3_2 --> logState3_3
        logState3_3 --> logState3_4
        logState3_4 --> logState3_5
    }

  FirstPage --> SecondPageTree : Serialization and\nDeserialization
  SecondPageTree --> SecondPage : Execution\nPreparation
```

> #### Serialization

Reviewer: Not mentioned here (although maybe we should): 
* What about the request context (like the query start time clock) - how is this serialized
* What about other request artefacts, such as `fetch_size` or maybe the `filter` - if provided.  How do these translate from request-to-request? 
* Does this work with post-request processing? 
* What about requests larger than the max window size of 10,000?

Author: - It is not supported yet, but actually it is just binary dumped into cursor
- It is not supported
- What do you mean?
- Good idea, I'll add

Reviewer: Anything that we know is not-supported should be mentioned.

    participant OpenSearchScrollRequest
    participant ContinuePageRequest

PlanSerializer ->>+ ProjectOperator : getPlanForSerialization
  ProjectOperator -->>- PlanSerializer : this
PlanSerializer ->>+ ProjectOperator : serialize
  Note over ProjectOperator : dump private fields
  ProjectOperator ->>+ ResourceMonitorPlan : getPlanForSerialization
    ResourceMonitorPlan -->>- ProjectOperator : delegate
  Note over ResourceMonitorPlan : ResourceMonitorPlan<br />is not serialized
  ProjectOperator ->>+ OpenSearchPagedIndexScan : serialize
    alt First page
      OpenSearchPagedIndexScan ->>+ OpenSearchScrollRequest : toCursor
        OpenSearchScrollRequest -->>- OpenSearchPagedIndexScan : serialized request
    else Subsequent page
      OpenSearchPagedIndexScan ->>+ ContinuePageRequest : toCursor
        ContinuePageRequest -->>- OpenSearchPagedIndexScan : serialized request
    end
    Note over OpenSearchPagedIndexScan : dump private fields
    OpenSearchPagedIndexScan -->>- ProjectOperator : serialized
  ProjectOperator -->>- PlanSerializer : serialized
Note over PlanSerializer : Zip to reduce size
```

#### Deserialization

> Deserialization restores previously serialized Physical Plan tree. The recovered tree is ready to execute and should return the next page of the search response. To complete the tree restoration, SQL engine should build a new request to the OpenSearch node. This request doesn't contain a search query, but it contains a search context reference -- `scrollID`. To create a new `ContinuePageRequest` object it is require to access to the instance of `OpenSearchStorageEngine`. `OpenSearchStorageEngine` can't be serialized and it exists as a singletone in the SQL plugin engine. `PlanSerializer` creates a customized deserialization binary object stream -- `CursorDeserializationStream`; this stream provides an interface to access to the `OpenSearchStorageEngine` object.

Reviewer: ```suggestion
Deserialization restores previously serialized Physical Plan tree. The recovered tree is ready to execute and should return the next page of the search response. To complete the tree restoration, SQL engine should build a new request to the OpenSearch node. This request doesn't contain a search query, but it contains a search context reference -- `scrollID`. To create a new `ContinuePageRequest` object it is require to access to the instance of `OpenSearchStorageEngine`. `OpenSearchStorageEngine` can't be serialized and it exists as a singleton in the SQL plugin engine. `PlanSerializer` creates a customized deserialization binary object stream -- `CursorDeserializationStream`. This stream provides an interface to access the `OpenSearchStorageEngine` object.
```

Reviewer: > This stream provides an interface to access the `OpenSearchStorageEngine` object.
I'm not sure what you mean by this?  Does this mean we can create a new `OpenSearchStorageEngine` object?

Author: No, we can't, it is a singleton.
